nilearn/mass_univariate/permuted_least_squares.py

```diff
@@ -440,92 +650,265 @@ def permuted_ols(
+    vfwe_scores_as_ranks = np.zeros((n_regressors, n_descriptors))
+    for vfwe_scores_as_ranks_part in vfwe_scores_as_ranks_parts:
+        vfwe_scores_as_ranks += vfwe_scores_as_ranks_part
+
+    vfwe_pvals = (n_perm + 1 - vfwe_scores_as_ranks) / float(1 + n_perm)
+
+    if threshold is not None:
+        # Cluster-size and cluster-mass FWE
+        csfwe_h0 = np.hstack((csfwe_h0_parts))
+        cmfwe_h0 = np.hstack((cmfwe_h0_parts))
+
+        csfwe_pvals = np.zeros_like(vfwe_pvals)
+        cmfwe_pvals = np.zeros_like(vfwe_pvals)
+        size_arr = np.zeros_like(vfwe_pvals).astype(int)
+        mass_arr = np.zeros_like(vfwe_pvals)
+
+        scores_original_data_4d = masker.inverse_transform(
+            scores_original_data.T
+        ).get_fdata()
+        bin_struct = ndimage.generate_binary_structure(3, 1)
+
+        for i_regressor in range(n_regressors):
+            scores_original_data_3d = scores_original_data_4d[..., i_regressor]
+
+            # Label the clusters for both cluster mass and size inference
+            labeled_arr3d, _ = ndimage.measurements.label(
+                scores_original_data_3d > threshold_t,
+                bin_struct,
+            )
+
+            if two_sided_test:
+                # Label positive and negative clusters separately
+                n_positive_clusters = np.max(labeled_arr3d)
+                temp_labeled_arr3d, _ = ndimage.measurements.label(
+                    scores_original_data_3d < -threshold_t,
+                    bin_struct,
+                )
+                temp_labeled_arr3d[
+                    temp_labeled_arr3d > threshold_t
+                ] += n_positive_clusters
+                labeled_arr3d = labeled_arr3d + temp_labeled_arr3d
+                del temp_labeled_arr3d
+
+            cluster_labels, idx, cluster_sizes = np.unique(
+                labeled_arr3d,
+                return_inverse=True,
+                return_counts=True,
+            )
+            assert cluster_labels[0] == 0  # the background
+
+            # Cluster mass-based inference
+            cluster_masses = np.zeros(cluster_labels.shape)
+            for j_val in cluster_labels[1:]:  # skip background
+                cluster_mass = np.sum(
+                    np.fabs(scores_original_data_3d[labeled_arr3d == j_val])
+                    - threshold_t
+                )
+                cluster_masses[j_val] = cluster_mass
+
+            p_cmfwe_vals = _null_to_p(
+                cluster_masses,
+                cmfwe_h0[i_regressor, :],
+                'larger',
+            )
+            p_cmfwe_map = p_cmfwe_vals[np.reshape(idx, labeled_arr3d.shape)]
+            mass_map = cluster_masses[np.reshape(idx, labeled_arr3d.shape)]
+
+            # Convert 3D to image, then to 1D
+            # There is a problem if the masker performs preprocessing,
+            # so we use apply_mask here.
+            cmfwe_pvals[i_regressor, :] = np.squeeze(
```

Comment:
There is a lot of similar code here.
I might be wrong, but I think we could factorize a bit with a for loop over something like `zip([cmfwe_pvals, csfwe_pvals], [mass_arr, size_arr])`. WDYT?